devel/query.lua

```diff
@@ -1,254 +1,507 @@
 --luacheck:skip-entirely
 local utils=require('utils')
 local validArgs = utils.invert({
  'help',
 
  'unit',
  'item',
+ 'plant',
+ 'building',
+ 'job',
+ 'workshopjob',
  'tile',
+ 'block',
  'table',
-
- 'query',
- 'querykeys',
- 'depth',
- 'keydepth',
- 'maxtablelength',
-
- 'listall',
- 'listfields',
- 'listkeys',
  'getfield',
- 'setkey',
 
- 'includeitall',
+ 'search',
+ 'findvalue',
+ 'maxdepth',
+ 'maxlength',
+ 'excludetype',
+ 'excludekind',
 
- 'safer',
+ 'noblacklist',
  'dumb',
+
+ 'showpaths',
+ 'setvalue',
+ 'oneline',
+ '1',
  'disableprint',
- 'debug'
+ 'debug',
+ 'debugdata'
 })
 local args = utils.processArgs({...}, validArgs)
-depth=nil
-keydepth=nil
-cur_depth = -1
-cur_keydepth = -1
-
-newkeyvalue=nil
-bprintfields=nil
-bprintkeys=nil
-space_field="   "
-fN=0
---kN=25
+local selection = nil
+local path_info = nil
+local path_info_pattern = nil
+local new_value = nil
+local find_value = nil
+local maxdepth = nil
+local cur_depth = -1
+local tilex = nil
+local tiley = nil
+local bool_flags = {}
+
 local help = [====[
 
 devel/query
 ===========
-Query is a script useful for finding and reading values of data structure fields.
-Purposes will likely be exclusive to writing lua script code.
-
-Fields: contents of tables
-Keys:   contents below non-table containers
-
-Keys and fields are essentially the same thing. The real difference is what
-code checks need to be in place for queries on keys vs fields. This is why
-they are differentiated.
-
-This is a recursive script which takes your selected data {table,unit,item,tile}
-and then iterates through it, then iterates through anything it finds. It does
-this recursively until it has walked over everything it is allowed. Everything
-it walks over it checks against any (optional) string/value queries, and if it
-finds a match it then prints it to the console.
-
-You can control most aspects of this process, the script is fairly flexible. So
-much so that you can easily create an infinitely recursing query and/or potentially
-crash Dwarf Fortress and DFHack. In previous iterations memory bloat was even a
-concern, where RAM would be used up in mere minutes or seconds; you can probably
-get this to happen as well if you are careless with the depth settings and don't
-print everything walked over (i.e. have a search term). The `kill-lua` command
-may be able to stop this script if it gets out of control.
-
-Before recursing or printing things to the console the script checks several things.
-A few important ones:
-
- - Is the data structure capable of being iterated?
- - Is the data structure pointing to a parent data structure?
- - Is the current level of recursion too high, and do we need to unwind it first?
- - Is the number of entries too high (eg. 70,000 table entries that would be printed)?
- - Is the data going to be usefully readable?
- - Does the field or key match the field or key query or queries?
- - Is printing fields allowed?
- - Is printing keys allowed?
```

Comment:
To be honest, I had written that because the script wasn't well organized. This section helped me rewrite it.

I'll re-add it though, with the necessary changes.

Reply:
Let me know if the new revised version is adequate